datalad/support/annexrepo.py

```diff
@@ -2937,15 +2937,15 @@ def commit(self, msg=None, options=None, _datalad_msg=False,
                     # Files which were staged but not among files
                     staged_not_to_commit = all_changed_staged.difference(files_changed_staged)
 
-                    if staged_not_to_commit or files_changed_notstaged:
+                    if files_changed_notstaged:
+                        self.add(files=list(files_changed_notstaged))
```

Comment:
uff... this whole changed/staged/etc hurts my brain ;-)  
I agree that probably a single condition (`if staged_not_to_commit`) is the most sensible to decide either we need a temp index or not.  But then I am getting lost in what needs to be reset and possible interactions from `adding` here which possibly could be not taken into account I think for `staged_not_to_commit` (since it differs with `files_changed_staged` so no `unstaged` before but staged now files are taken into account).

By doing this `.add` here we in effect stage `files_changed_notstaged`.  So I wonder if we could/should move assignments for `files_changed_staged` and `staged_not_to_commit` after this add?:

```patch
diff --git a/datalad/support/annexrepo.py b/datalad/support/annexrepo.py
index 449fe297f..e954125b3 100644
--- a/datalad/support/annexrepo.py
+++ b/datalad/support/annexrepo.py
@@ -2919,27 +2919,29 @@ class AnnexRepo(GitRepo, RepoInterface):
                     # files "jumped" between git/annex.  Then also preparing a
                     # custom index and calling "commit" without files resolves
                     # the issue
-                    all_changed_staged = \
-                        set(self.get_changed_files(staged=True))
-
                     files_normalized = [
                         _normalize_path(self.path, f) if isabs(f) else f
                         for f in files
                     ]
 
-                    files_changed_staged = \
-                        set(self.get_changed_files(staged=True, files=files_normalized))
                     files_changed_notstaged = \
                         set() \
                         if direct_mode \
                         else set(self.get_changed_files(staged=False, files=files_normalized))
 
-                    # Files which were staged but not among files
-                    staged_not_to_commit = all_changed_staged.difference(files_changed_staged)
-
+                    # Stage all files which are not staged yet
                     if files_changed_notstaged:
                         self.add(files=list(files_changed_notstaged))
 
+                    files_changed_staged = \
+                        set(self.get_changed_files(staged=True, files=files_normalized))
+
+                    all_changed_staged = \
+                        set(self.get_changed_files(staged=True))
+
+                    # Files which were staged but not among files
+                    staged_not_to_commit = all_changed_staged.difference(files_changed_staged)
+
                     if staged_not_to_commit:
                         # Need an alternative index_file
                         with make_tempfile(dir=opj(self.path,
```
or am i ruining some logic (relying on original values before any changes) this way?

Reply:
> uff... this whole changed/staged/etc hurts my brain ;-)

Mine too :]

> By doing this `.add` here we in effect stage `files_changed_notstaged`.

Yes, though functionally it is no different then doing it in the old spot.  We just avoid the unnecessary temporary index.

> So I wonder if we could/should move assignments for `files_changed_staged` and `staged_not_to_commit` after this add?

There are two places where the values of `files_changed_{not,}staged` come in: (1) determining the value of `staged_not_to_commit` and (2) doing the clean-up reset of the final index.  No. 1 doesn't change based on whether `files_changed_staged` and `all_changed_staged` are updated to include the just added `files_changed_notstaged` because that is adding files to both sets, which will be dropped with the `all_changed_staged - files_changed_notstaged` difference.  No 2. isn't affected because the union of `files_changed_notstaged` and `files_changed_staged` are used for the reset, so how we partition files among the two sets doesn't matter.

Reply:
ok, let's then stick to your PR as is, thanks!